server/src/main/java/org/elasticsearch/common/logging/DynamicContextDataProvider.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.core.util.ContextDataProvider;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An implementation of log4j2's {@link ContextDataProvider} that can be configured at runtime
+ * (after being loaded by log4j's init mechanism).
+ *
+ * A {@link ContextDataProvider} can enrich a logging event with additional "context data" (a string-to-string map of additional fields).
+ * Elasticsearch (via {@link ECSJsonLayout} logs these additional fields in the json formatted log.
+ *
+ * Log4j loads {@link ContextDataProvider} instances using SPI during logging initialisation. In Elasticsearch this initialisation happens
+ * early in the bootstrap process, before Elasticsearch modules and plugins are loaded. Thus, it is not possible for an Elasticsearch plugin
+ * to register a {@link ContextDataProvider} service with log4j.
+ *
+ * Instead, Elasticsearch allows plugins to implement the {@link LoggingDataProvider} which is loaded via Elasticsearch's SPI implementation
+ * (as part of plugin loading). This {@link DynamicContextDataProvider} class is the bridge between the Elasticsearch and log4j systems.
+ */
+public class DynamicContextDataProvider implements ContextDataProvider {
+
+    // This is not a set-once because some integration tests may try to set it twice
+    private static final AtomicReference<List<? extends LoggingDataProvider>> DATA_PROVIDERS = new AtomicReference<>();
+
+    public static void setDataProviders(List<? extends LoggingDataProvider> dataProviders) {
+        DynamicContextDataProvider.DATA_PROVIDERS.compareAndSet(null, List.copyOf(dataProviders));
+    }
+
+    @Override
+    public Map<String, String> supplyContextData() {
```

Comment:
As I understand it this is called on every logging call. Is there any possibility to cache the generated map?

Reply:
Assuming we want this system to be able to log things that change (and the logging system doesn't know when they change) such as active username / trace-id, etc (which is why I want it), then caching is tricky.

Some options that might work
1. Always reuse the last map, and make each provider responsible for _clearing_ any keys that ought not to be set, as well as populating keys that should be. I'm not sure about the thread safety of that, but we can experiment if needed. It has the risk that we might log something inaccurate if one of the providers fails to clear a field.
2. As above (1), but set all the values to null before calling the providers. That increases safety at the expense of CPU time.
3. Create the new map with the same keys as the last map. Like (2) but easier to make thread safe, but still has high object churn (with fewer reallocations)
4. Create the new map with the same size as the last map. That would reduce internal reallocations in the map, like (3) but not really cache anything.
5. Use a custom `Map` implementation (or a custom `StringMap` which is what log4j really uses here), with one of the following behaviours:
   - Have each provider return their own map, and our custom map is just a _merging view_ over them. That means providers that have constant data can return a constant map
   - Extend the `LoggingDataProvider` so that it returns the set of keys as constants (a new method) and we can define a map implementation that takes advantage of the known set of keys to reduce memory usage/churn.

I suspect having a predicted size and allocating the map to be that size is the only option that is
- simple enough to implement
- thread safe
- sufficiently performant to justify

Reply:
I implemented version that uses a fixed-keys map here https://github.com/elastic/elasticsearch/commit/8499a532e1923f6c0b6affecee45eee91e4abf03
I haven't actually tested the performance gains that might come from that, but it seems like a viable option if we want to pursue it.

Reply:
thinking out loud..
can we instead of having `DynamicContextDataProvider` implement a `ContextDataProvider` directly in every plugin?
do we have to aggregate `LoggingDataProvider` in` DynamicContextDatProvider`? it feels like log4j is already doing this as it expects multiple `ContextDataProviders`

Reply:
I don't think there's any way to do that.
Log4j won't load `ContextDataProvider` services from our plugins because they aren't on the classpath when we initialise log4j, and as far as I know, there is no way to reinitialise the services after we've loaded plugins & modules.
We some sort of bridge from what log4j can see at init-time into our module structure.

Reply:
@rjernst Did you want to explore any of the other possible caching options?

I've implemented a cached "largest previous map size" so that we can allocate the new map to be the same size, but we could explore options to do more than that if you would like.